lib/scholar/metrics.ex

```diff
@@ -266,71 +246,101 @@ defmodule Scholar.Metrics do
 
   ## Options
 
-    * `:threshold` - threshold for truth value of the predictions.
-      Defaults to `0.5`
+    * `:num_classes` - Number of classes contained in the input tensors
 
   ## Examples
 
-      iex> Scholar.Metrics.sensitivity(Nx.tensor([0, 1, 1, 1]), Nx.tensor([1, 0, 1, 1]))
+      iex> y_true = Nx.tensor([0, 1, 1, 1, 1, 0, 2, 1, 0, 1], type: {:u, 32})
+      iex> y_pred = Nx.tensor([0, 2, 1, 1, 2, 2, 2, 0, 0, 1], type: {:u, 32})
+      iex> Scholar.Metrics.sensitivity(y_true, y_pred, num_classes: 3)
       #Nx.Tensor<
-        f32
-        0.6666666865348816
+        f32[3]
+        [0.6666666865348816, 0.5, 1.0]
       >
 
   """
   defn sensitivity(y_true, y_pred, opts \\ []) do
-    assert_shape(y_true, Nx.shape(y_pred))
-
-    opts = keyword!(opts, threshold: 0.5)
+    opts = keyword!(opts, [:num_classes])
+    assert_shape_pattern(y_true, {_})
+    assert_shape(y_pred, Nx.shape(y_true))
 
     recall(y_true, y_pred, opts)
   end
 
   @doc ~S"""
-  Computes the specificity of the given predictions
-  with respect to the given targets.
+  Computes the specificity of the given predictions with respect
+  to the given targets for binary classification problems.
 
   ## Argument Shapes
 
     * `y_true` - $\(d_0, d_1, ..., d_n\)$
     * `y_pred` - $\(d_0, d_1, ..., d_n\)$
 
-  ## Options
-
-    * `:threshold` - threshold for truth value of the predictions.
-      Defaults to `0.5`
-
   ## Examples
 
-      iex> Scholar.Metrics.specificity(Nx.tensor([0, 1, 1, 1]), Nx.tensor([1, 0, 1, 1]))
+      iex> Scholar.Metrics.binary_specificity(Nx.tensor([0, 1, 1, 1]), Nx.tensor([1, 0, 1, 1]))
       #Nx.Tensor<
         f32
         0.0
       >
 
   """
-  defn specificity(y_true, y_pred, opts \\ []) do
+  defn binary_specificity(y_true, y_pred) do
     assert_shape(y_true, Nx.shape(y_pred))
 
-    opts = keyword!(opts, threshold: 0.5)
-
-    thresholded_preds = Nx.greater(y_pred, opts[:threshold])
-
     true_negatives =
-      thresholded_preds
+      y_pred
       |> Nx.equal(y_true)
-      |> Nx.logical_and(Nx.equal(thresholded_preds, 0))
+      |> Nx.logical_and(Nx.equal(y_pred, 0))
       |> Nx.sum()
 
     false_positives =
-      thresholded_preds
+      y_pred
       |> Nx.not_equal(y_true)
-      |> Nx.logical_and(Nx.equal(thresholded_preds, 1))
+      |> Nx.logical_and(Nx.equal(y_pred, 1))
       |> Nx.sum()
 
     Nx.divide(true_negatives, false_positives + true_negatives + 1.0e-16)
   end
 
+  @doc ~S"""
+  Computes the specificity of the given predictions with respect
+  to the given targets for multi-class classification problems.
+
+  ## Argument Shapes
+
+    * `y_true` - $\(d_0, d_1, ..., d_n\)$
+    * `y_pred` - $\(d_0, d_1, ..., d_n\)$
```

Comment:
These are always 1-d tensors, no? We need to adjust in other places too.

Reply:
Yeah, they should be 1-d tensors as we are only dealing with single-label problems. For multilabel problems, we probably should have a different set of functions?

Reply:
Yeah, I'm not sure about multi labels, we can investigate that separately.